File: src/main/java/supermetroid/trackable/BooleanTrackable.java
package supermetroid.trackable;

import supermetroid.MajorTracker;
import supermetroid.State;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.util.Objects;
import java.util.function.Function;

public class BooleanTrackable extends Trackable {
  private ImageIcon on;
  private ImageIcon off;
  private Function<State, Boolean> booleanMemoryReader;
  private Boolean cache;

  public BooleanTrackable(String image, Function<State, Boolean> booleanMemoryReader) {
    try {
      this.on = new ImageIcon(ImageIO.read(BooleanTrackable.class.getResourceAsStream("/on/" + image + ".png")));
      this.off = new ImageIcon(ImageIO.read(BooleanTrackable.class.getResourceAsStream("/off/" + image + ".png")));
    } catch (Exception e) {
      System.out.println("Error loading boolean trackable image for " + image);
      e.printStackTrace();
    }

    this.booleanMemoryReader = booleanMemoryReader;
  }

  @Override
  public void setIcon(State state) {
    if (cache == null || cache != booleanMemoryReader.apply(state)) {
      cache = booleanMemoryReader.apply(state);
      label.setIcon(cache ? on : off);
    }
  }

  @Override
  public void setDefault() {
    label.setIcon(off);
  }

  public static final BooleanTrackable CHARGE = new BooleanTrackable("charge", state -> (state.charge & 0x10) > 0);
  public static final BooleanTrackable ICE = new BooleanTrackable("ice", state -> (state.beams & 0x2) > 0);
  public static final BooleanTrackable WAVE = new BooleanTrackable("wave", state -> (state.beams & 0x1) > 0);
  public static final BooleanTrackable SPAZER = new BooleanTrackable("spazer", state -> (state.beams & 0x4) > 0);
  public static final BooleanTrackable PLASMA = new BooleanTrackable("plasma", state -> (state.beams & 0x8) > 0);

  public static final BooleanTrackable MORPH = new BooleanTrackable("morph", state -> (state.equipmentA & 0x4) > 0);
  public static final BooleanTrackable VARIA = new BooleanTrackable("varia", state -> (state.equipmentA & 0x1) > 0);
  public static final BooleanTrackable SPRING_BALL = new BooleanTrackable("spring-ball", state -> (state.equipmentA & 0x2) > 0);
  public static final BooleanTrackable HI_JUMP = new BooleanTrackable("hi-jump", state -> (state.equipmentB & 0x1) > 0);
  public static final BooleanTrackable SPACE_JUMP = new BooleanTrackable("space-jump", state -> (state.equipmentB & 0x2) > 0);
  public static final BooleanTrackable BOMBS = new BooleanTrackable("bombs", state -> (state.equipmentB & 0x10) > 0);
  public static final BooleanTrackable GRAVITY = new BooleanTrackable("gravity", state -> (state.equipmentA & 0x20) > 0);
  public static final BooleanTrackable SPEED_BOOSTER = new BooleanTrackable("speed-booster", state -> (state.equipmentB & 0x20) > 0);
  public static final BooleanTrackable SCREW_ATTACK = new BooleanTrackable("screw-attack", state -> (state.equipmentA & 0x8) > 0);

  public static final BooleanTrackable PHANTOON = new BooleanTrackable("phantoon", state -> (state.wreckedShipBosses & 0x1) > 0);
  public static final BooleanTrackable RIDLEY = new BooleanTrackable("ridley", state -> (state.norfairBosses & 0x1) > 0);
  public static final BooleanTrackable KRAID = new BooleanTrackable("kraid", state -> (state.brinstarBosses & 0x1) > 0);
  public static final BooleanTrackable DRAYGON = new BooleanTrackable("draygon", state -> (state.maridiaBosses & 0x1) > 0);

  public static final BooleanTrackable CROCOMIRE = new BooleanTrackable("crocomire", state -> (state.norfairBosses & 0x2) > 0);
  public static final BooleanTrackable SHAKTOOL = new BooleanTrackable("shaktool", state -> (state.events & 0x20) > 0);
}

File: src/main/java/supermetroid/trackable/BooleanTrackable.kt
package supermetroid.trackable

import supermetroid.State
import supermetroid.readResourceToIcon
import java.util.function.Function
import javax.swing.ImageIcon

class BooleanTrackable(image: String, private val booleanMemoryReader: Function<State, Boolean>) : Trackable() {
    private var on: ImageIcon = readResourceToIcon("/on/$image.png")
    private var off: ImageIcon = readResourceToIcon("/off/$image.png")
    private var current = false

    override fun setIcon(state: State) {
        if (current != booleanMemoryReader.apply(state)) {
            current = booleanMemoryReader.apply(state)
            label.icon = if (current) on else off
        }
    }

    override fun setDefault() {
        label.icon = off
    }

    companion object {
        val CHARGE: BooleanTrackable = BooleanTrackable("charge") { state: State -> (state.charge and 0x10) > 0 }
        val ICE: BooleanTrackable = BooleanTrackable("ice") { state: State -> (state.beams and 0x2) > 0 }
        val WAVE: BooleanTrackable = BooleanTrackable("wave") { state: State -> (state.beams and 0x1) > 0 }
        val SPAZER: BooleanTrackable = BooleanTrackable("spazer") { state: State -> (state.beams and 0x4) > 0 }
        val PLASMA: BooleanTrackable = BooleanTrackable("plasma") { state: State -> (state.beams and 0x8) > 0 }

        val MORPH: BooleanTrackable = BooleanTrackable("morph") { state: State -> (state.equipmentA and 0x4) > 0 }
        val VARIA: BooleanTrackable = BooleanTrackable("varia") { state: State -> (state.equipmentA and 0x1) > 0 }
        val SPRING_BALL: BooleanTrackable = BooleanTrackable("spring-ball") { state: State -> (state.equipmentA and 0x2) > 0 }
        val HI_JUMP: BooleanTrackable = BooleanTrackable("hi-jump") { state: State -> (state.equipmentB and 0x1) > 0 }
        val SPACE_JUMP: BooleanTrackable = BooleanTrackable("space-jump") { state: State -> (state.equipmentB and 0x2) > 0 }
        val BOMBS: BooleanTrackable = BooleanTrackable("bombs") { state: State -> (state.equipmentB and 0x10) > 0 }
        val GRAVITY: BooleanTrackable = BooleanTrackable("gravity") { state: State -> (state.equipmentA and 0x20) > 0 }
        val SPEED_BOOSTER: BooleanTrackable = BooleanTrackable("speed-booster") { state: State -> (state.equipmentB and 0x20) > 0 }
        val SCREW_ATTACK: BooleanTrackable = BooleanTrackable("screw-attack") { state: State -> (state.equipmentA and 0x8) > 0 }

        val PHANTOON: BooleanTrackable = BooleanTrackable("phantoon") { state: State -> (state.wreckedShipBosses and 0x1) > 0 }
        val RIDLEY: BooleanTrackable = BooleanTrackable("ridley") { state: State -> (state.norfairBosses and 0x1) > 0 }
        val KRAID: BooleanTrackable = BooleanTrackable("kraid") { state: State -> (state.brinstarBosses and 0x1) > 0 }
        val DRAYGON: BooleanTrackable = BooleanTrackable("draygon") { state: State -> (state.maridiaBosses and 0x1) > 0 }

        val CROCOMIRE: BooleanTrackable = BooleanTrackable("crocomire") { state: State -> (state.norfairBosses and 0x2) > 0 }
        val SHAKTOOL: BooleanTrackable = BooleanTrackable("shaktool") { state: State -> (state.events and 0x20) > 0 }
    }
}
